src/main/java/com/marklogic/spark/reader/optic/OpticReadContext.java

```diff
@@ -53,12 +54,44 @@ public class OpticReadContext extends ContextSupport {
     private StructType schema;
     private long serverTimestamp;
     private List<OpticFilter> opticFilters;
+    private final Map<String, String> userDefinedParams;
+    private final Map<String, String> columnParamDefaults;
+    private Map<String, String> columnParamValues;
     private final long batchSize;
 
     public OpticReadContext(Map<String, String> properties, StructType schema, int defaultMinPartitions) {
         super(properties);
         this.schema = schema;
 
+
+        // Stash a map of the user-defined parameters (if any) so that they aren't calculated on every call.
```

Comment:
A clean-code comment - this constructor was already a little complex. To keep the code more readable, all this new stuff should go into a new private method that initializes the different maps.